### ISSUE TITLE
benchmark large datasets

### ISSUE THREAD

**Issue by thibautjombart** [thibautjombart/treescape#47]
I have just tried with this dataset:
http://beast-mcmc.googlecode.com/files/Influenza%20BEAST%20Practical.zip

and it either froze or is real slow. We need to benchmark the method and see how it behaves. We might want to set default limits in tree size and/or numbers of trees.

**Comment by MichelleKendall** [comment]
Good idea. We could even allow the user to upload a set of trees and pick a value x, so that Shiny takes a random sample of x trees (from the second half of the posterior?) to compare?

**Comment by thibautjombart** [comment]
yep

**Comment by jalmagro** [comment]
I would assume that R is eating all the memory, is that right? We may consider using something like big.matrix in the future to handle this (it consumes way less memory than the R native types).

**Comment by thibautjombart** [comment]
I am not sure this is the problem. Time is clearly not taken by memory allocation for the basic matrix of pairwise comparisons:
```r
> system.time(D<-dist(1:1e4))
   user  system elapsed 
  0.579   0.102   0.681 
> dim(as.matrix(D))
[1] 10000 10000
> print(object.size(D), unit="auto")
381.4 Mb
> 
```
Though I haven't looked at the details of the algo, we would need many copies of the data and intermediate objects for RAM to be a problem. It was definitely not the case on my 64 bits desktop with 16GB RAM.

**Comment by jalmagro** [comment]
Oh I see, I was thinking about the multiDist function but even in that case you should have enough memory. For 10K trees with 165 tips, if the save memory option is not enabled, the method will create a matrix of size 10000 * (choose(165,2) + 165) numeric elements but that should be "only" around 1GB in R, if I'm not mistaken. Anyway, pairwise distances for 10K samples is around 50M metric vector comparisons so the inner loop (although vectorized) is going to be very slow in R. For very large datasets we'll have to move the whole method to RCPP space and use multiple cores (chunking the matrix). I might end up doing that in the near future, I'll keep you posted.

**Comment by MichelleKendall** [comment]
Feedback from users was that they often just get a blank screen in Shiny after uploading their trees, which, combined with red error messages, wasn't very comforting! Obviously there will be a more elegant solution and better ways of handling large datasets but for now I've just made Shiny output a warning message (in the R window) when num_trees * num_tips > 500, saying that Shiny might be slow. I've also removed those annoying "Pick lambda in [0,1]" messages. One error message remains, which I'm looking into:
```{r}
Error in FUN(X[[i]], ...) : 
  only defined on a data frame with all numeric variables
```

**Comment by MichelleKendall** [comment]
I have removed the error message - it was a circular problem when choosing the number of axes retained. For large datasets I've introduced a better though not perfect solution: I have set it so that Shiny automatically takes a small random sample of the trees uploaded and plots them, and the user can THEN increase the number (or choose not to sample at all). This prevents error messages and makes it a bit more understandable when it does hang for ages. I've also added a "loading" progress bar.
